renku/core/models/entity.py

```diff
@@ -0,0 +1,121 @@
+# limitations under the License.
+"""Represent provenance entities."""
+
+from pathlib import Path
+from typing import List, Union
+from urllib.parse import quote
+
+from renku.core.incubation.database import Persistent
+from renku.core.models.calamus import JsonLDSchema, Nested, fields, prov, renku, wfprov
+from renku.core.utils.git import get_object_hash
+from renku.core.utils.urls import get_host
+
+
+class Entity(Persistent):
+    """Represent a file."""
+
+    def __init__(self, id: str, checksum: str, path: Union[Path, str]):
+        self.id: str = id
+        self.path: Path = path
+        self.checksum: str = checksum
+
+    @staticmethod
+    def generate_id(hostname: str, checksum: str, path: Union[Path, str]) -> str:
+        """Generate an Entity identifier."""
+        quoted_path = quote(str(path).strip("/"))
+
+        return f"https://{hostname}/entities/{checksum}/{quoted_path}"
+
+    @classmethod
+    def from_revision(cls, client, path: Union[Path, str], revision: str = "HEAD", find_previous: bool = True):
+        """Return dependency from given path and revision."""
+        if find_previous:
+            revision = client.find_previous_commit(path, revision=revision)
+
+        client, commit, path = client.resolve_in_submodules(revision, path,)
+
+        hostname = get_host(client)
+        checksum = get_object_hash(repo=client.repo, revision=revision, path=path)
+        # TODO: What if checksum is None
+        # TODO: What would be checksum for a directory if it's not committed yet.
+        id = cls.generate_id(hostname=hostname, checksum=checksum, path=path)
+
+        absolute_path = client.path / path
+        if str(path) != "." and absolute_path.is_dir():
+            files_in_commit = commit.stats.files
+            members: List[Entity] = []
+
+            for member in absolute_path.iterdir():
+                if member.name == ".gitkeep":
+                    continue
+
+                member_path = str(member.relative_to(client.path))
+                find_previous = True
+
+                if member_path in files_in_commit:
+                    # we already know the newest commit, no need to look it up
+                    find_previous = False
+
+                try:
+                    assert all(member_path != m.path for m in members)
+
+                    members.append(cls.from_revision(client, member_path, commit, find_previous=find_previous))
+                except KeyError:
+                    pass
+
+            entity = Collection(id=id, checksum=checksum, path=path, members=[])
+        else:
+            entity = cls(id=id, checksum=checksum, path=path)
+
+        return entity
+
+
+class Collection(Entity):
+    """Represent a directory with files."""
+
+    def __init__(self, id: str, checksum: str, path: Union[Path, str], members: List[Entity]):
+        super().__init__(id, checksum, path)
+        self.members: List[Entity] = members
```

Comment:
do we need to switch lists on `Persistent` classes over to `PersistentList` ?

Reply:
I believe for lists it does not matter since they don't call `__hash__` or `__eq__` when an object is added. For dict and set we probably need to check.

Reply:
I was more thinking of 
```
Note that mutating a non-persistent attribute of a persistent object such as a dict or list will not cause the containing object to be changed. Instead you can either explicitly control the state as described below, or use a PersistentList or PersistentMapping.
```
from https://persistent.readthedocs.io/en/latest/using.html#associating-an-object-with-a-data-manager

So as I understand it, adding a member wouldn't cause the `_p_changed` of the `Collection` to be updated.

Reply:
Actually, if we change a Collection then its checksum changes and we need to create a new object with update `oid` and then replace the current object with it. In that case, we don't need to change the current object.